crates/bevy_ecs/src/schedule/system_graph.rs

```diff
@@ -0,0 +1,264 @@
+use crate::schedule::{SystemDescriptor, SystemLabel, SystemSet};
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    sync::{
+        atomic::{AtomicU32, Ordering},
+        Arc, Mutex,
+    },
+};
+
+static NEXT_NODE_ID: AtomicU32 = AtomicU32::new(0);
+
+/// A builder for creating graphs of dependent parallel execution within a `SystemStage`.
+///
+/// # Sequential Execution
+/// The simplest graph is a sequence of systems that need to run one after another.
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// let graph = SystemGraph::new();
+/// graph
+///   .root(sys_a.system())
+///   .then(sys_b.system())
+///   .then(sys_c.system());
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+///
+/// # Fan Out
+/// `SystemGraphNode::then` can be called repeatedly on the same node to create multiple fan-out
+/// branches. All fanned out systems will not execute until the original has finished.
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// # fn sys_d() {}
+/// # fn sys_e() {}
+/// let graph = SystemGraph::new();
+///
+/// let start_a = graph.root(sys_a.system());
+///
+/// start_a.then(sys_b.system());
+/// start_a.then(sys_c.system());
+///
+/// start_a
+///     .then(sys_d.system())
+///     .then(sys_e.system());
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+///
+/// # Fan In
+/// A graph node can wait on multiple systems before running.
+/// `SystemGraphJoinExt::join_into` is implemented on any type that iterates over
+/// `SystemGraphNode`.
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # fn sys_a() {}
+/// # fn sys_b() {}
+/// # fn sys_c() {}
+/// # fn sys_d() {}
+/// # fn sys_e() {}
+/// let graph = SystemGraph::new();
+///
+/// let start_a = graph.root(sys_a.system());
+/// let start_b = graph.root(sys_b.system());
+/// let start_c = graph.root(sys_c.system());
+///
+/// vec![start_a, start_b, start_c]
+///     .join_into(sys_d.system())
+///     .then(sys_e.system());
+///
+/// // Convert into a SystemSet
+/// let system_set: SystemSet = graph.into();
+/// ```
+#[derive(Clone, Default)]
+pub struct SystemGraph(Arc<Mutex<HashMap<NodeId, SystemDescriptor>>>);
+
+impl SystemGraph {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn root(&self, system: impl Into<SystemDescriptor>) -> SystemGraphNode {
```

Comment:
Any specific reason we have this take a `&self`?
This probably jumps back to the Arc/Mutex thing I mentioned before. 
However, because this is a "builder"-like struct, it would make more sense for this to take `&mut self` or `self`

Reply:
It takes &self since it all the state is interior to it. I guess &mut self might make sense for enforcing users mark variables as mut, but it's not strictly necessary. Using self would not be wise either since you can have multiple roots in a given graph.